pkg/controllers/customeradmin/group_mapping.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/sirupsen/logrus"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func setIfDifferent(field *string, value string) bool {
+	changed := (*field != value)
+	*field = value
+	return changed
+}
+
+func setLabelIfDifferent(labels map[string]string, key string, value string) bool {
+	current, exists := labels[key]
+	labels[key] = value
+	return !exists || current != value
+}
+
+func fromMSGraphGroup(log *logrus.Entry, kubeGroup *v1.Group, group *graphrbac.ADGroup, groupName string, msGroupMembers []graphrbac.AADObject) (*v1.Group, bool) {
+	changed := false
+	g := kubeGroup
+	if g == nil {
+		g = &v1.Group{
+			ObjectMeta: meta_v1.ObjectMeta{
+				Name: groupName,
+			},
+			Users: []string{},
+		}
+		changed = true
+	}
+	if g.ObjectMeta.Labels == nil {
+		g.ObjectMeta.Labels = map[string]string{}
+		changed = true
+	}
+	if group != nil {
+		if setLabelIfDifferent(g.ObjectMeta.Labels, "ADGroup.ObjectID", *group.ObjectID) {
+			changed = true
+		}
+		if setLabelIfDifferent(g.ObjectMeta.Labels, "ADGroup.DisplayName", *group.DisplayName) {
+			changed = true
+		}
+	}
+	kusers := map[string]struct{}{}
+	for _, user := range g.Users {
+		kusers[user] = struct{}{}
+	}
+	if len(msGroupMembers) == 0 && len(g.Users) > 0 {
+		g.Users = []string{}
+		changed = true
+	}
+	for _, user := range msGroupMembers {
+		if _, ok := kusers[*user.UserPrincipalName]; !ok {
+			g.Users = append(g.Users, *user.UserPrincipalName)
+			log.Debugf("adding %s to %s", *user.UserPrincipalName, kubeGroup.Name)
+			changed = true
+		}
+	}
+	return g, changed
+}
+
+// https://docs.microsoft.com/en-us/graph/api/group-list
+// To list AAD groups, this code needs to have the clientID
+// of an application with the following permissions:
+// API: Microsoft Graph
+//   Application permissions:
+//      Read all groups
+func newAADGroupsClient(aadConfig map[string]string) (*graphrbac.GroupsClient, error) {
+	c := auth.NewClientCredentialsConfig(aadConfig["aadClientId"], aadConfig["aadClientSecret"], aadConfig["tenantId"])
+	c.Resource = azure.PublicCloud.GraphEndpoint
+
+	authorizer, err := c.Authorizer()
+	if err != nil {
+		return nil, err
+	}
+
+	gc := graphrbac.NewGroupsClient(aadConfig["tenantId"])
+	gc.Authorizer = authorizer
+	return &gc, nil
+}
+
+func getAADGroupMembers(gc *graphrbac.GroupsClient, groupID string) ([]graphrbac.AADObject, error) {
+	users, err := gc.GetGroupMembers(context.Background(), groupID)
+	if err != nil {
+		return nil, err
+	}
+	members := []graphrbac.AADObject{}
+	for users.NotDone() {
+		for _, user := range users.Values() {
+			members = append(members, user)
+		}
+
+		if err = users.Next(); err != nil {
+			return nil, err
+		}
+	}
+	return members, nil
+}
+
+func getAADGroup(gc *graphrbac.GroupsClient, name string) (*graphrbac.ADGroup, error) {
+	// https://msdn.microsoft.com/en-gb/library/azure/ad/graph/api/groups-operations
+	groups, err := gc.List(context.Background(), fmt.Sprintf("displayName eq '%s'", name))
```

Comment:
I think we're going to get in hot water fast if we do this.  I think we have to accept group GUIDs in the external API and go straight to GetGroupMembers.

Reply:
ok